File: src/main/java/lmller/github/io/gofkotlin/prototype/Prototype.java
package lmller.github.io.gofkotlin.prototype;

public class Prototype {
    public static class EMail {

        private String recipient;
        private String subject;

        private String message;

        public EMail(String recipient, String subject, String message) {
            this.recipient = recipient;
            this.subject = subject;
            this.message = message;
        }

        private EMail(EMail src){
            this.recipient = src.recipient;
            this.subject = src.subject;
            this.message = src.message;
        }


        public EMail copy() {
            return new EMail(this);
        }
    }

    public static void main(String[] args) {
        EMail mail = new EMail("abc@example.com", "Hello", "Don't know what to write.");

        EMail copy = mail.copy();

        copy.recipient = "other@example.com";

        System.out.println("Email1 goes to " + mail.recipient + " with subject " + mail.subject);
        System.out.println("Email2 goes to " + copy.recipient + " with subject " + copy.subject);
    }
}

File: src/main/java/lmller/github/io/gofkotlin/prototype/Prototype.kt
package lmller.github.io.gofkotlin.prototype

data class EMail(var recipient: String, var subject: String?, var message: String?) {
    fun quote(startIndex: Int) : String {
        return "> ${message?.substring(startIndex)}"
    }
}

fun main(args: Array<String>) {
    val mail = EMail("abc@example.com", "Hello", "Don't know what to write.")

    val copy = mail.copy(recipient = "other@example.com")

    println("Email1 goes to " + mail.recipient + " with subject " + mail.subject)
    println("Email2 goes to " + copy.recipient + " with subject " + copy.subject)
}